Allows posting to prom rules endpoints via ds_proxy

owen-d opened the pull request:
This allows using the datasource proxy to create rule groups in Cortex via: https://github.com/cortexproject/cortex/blob/master/pkg/api/api.go#L301

owen-d committed:
allows posting to prom rules endpoints via ds_proxy

domasx2 commented:
perhaps also check that user is Editor or Admin?

owen-d commented:
That makes sense, but I think there'd be a lot of if statements involved. Is there a reason we're doing this logic here rather than the `plugin.json` file's `routes` field?

domasx2 commented:
> That makes sense, but I think there'd be a lot of if statements involved. Is there a reason we're doing this logic here rather than the `plugin.json` file's `routes` field?

As a precaution I guess, routes are not configured by default for the prometheus plugin afaik. Anyway I don't feel strongly, just a suggestion

owen-d commented:
> As a precaution I guess, routes are not configured by default for the prometheus plugin afaik.

I'll try to move the routing logic to the plugin's definition in order to be more idiomatic and reduce reimplementing logic and see how that works out :+1:

papagian reviewed (commented):
I think we need the same also for DELETE requests

owen-d committed:
prom proxy routes via plugin and fix proxy route matching bug

owen-d commented:
Edit: I found and fixed a bug (test included) where routes were being validated incorrectly. I've also moved the routing logic for the prometheus proxy to the plugin file, which is more idiomatic.

papagian commented on pkg/api/pluginproxy/ds_proxy_test.go:
```diff
@@ -895,3 +895,40 @@ func runDatasourceAuthTest(t *testing.T, test *testCase) {
+
+func Test_PathCheck(t *testing.T) {
+	// Ensure that we test routes appropriately. This test reproduces a historical bug where two routes were defined with different role requirements but the same method and the more privileged route was tested first. Here we ensure auth checks are applied based on the correct route, not just the method.
+	plugin := &plugins.DataSourcePlugin{
```
Another interesting case (if even possible) would be if the routes had the same path as well.

owen-d replied:
I think two route entries with the same method + same path are indistinguishable and so that'd be considered a schema bug I believe. We could probably validate that during plugin loading, but it seems out of scope for this PR.

papagian reviewed (approved):
LGTM, I really like this solution

owen-d committed:
bump ci
